### ISSUE TITLE
Summary of differences between this and react-three?

### ISSUE THREAD

**Issue by Izzimach** [toxicFork/react-three-renderer#28]
The question was brought up here https://github.com/Izzimach/react-three/issues/78 about differences between this project and [react-three](https://github.com/Izzimach/react-three).

I haven't looked at this code enough to make an informed summary. I would also like to add something in react-three's README saying "if you need XXX, you should use react-three-renderer instead"

Can you comment on the differences and why a person would choose one or other?

Thanks and cool to see this project!

**Comment by toxicFork** [comment]
Hi @Izzimach :)

TLDR: `react-three-renderer` is aiming to be the `react-native` for `THREE.js`.

# High level:

Both `react-three-renderer` and `react-three` still need to have everything wrapped with the whole `<Renderer>`.

React-three-renderer creates `native` components, where you can just do:
```jsx
return <mesh><geometry/><material/></mesh>;
```
instead of:
```jsx
var RTR = require('react-three-renderer');
var Mesh = RTR.Mesh;
var Geometry = RTR.Geometry;
var Material = RTR.Material;

return (<Mesh><Geometry/><Material/></Mesh>);
```

However you can still use composite components within natives, for example:

```jsx
<scene>
  <Car/>
  <Road/>
  <Tree/>
</scene>
```
And Car's render would look like:
```jsx
// Car.js
var Car = React.createClass({
  render: function(){
    return (<mesh><geometry ... /><material ... /></mesh>);
  }
});
```

So it is a bit less effort to use `react-three-renderer`
It means it is easier to extend `react-three` than `react-three-renderer` if for example the new `THREE.js` added `pyramid geometry`. However I am working on making that a bit easier for `react-three-renderer` :)

For fun, [there is an example of using `react-three-renderer` without using `react-dom`](https://github.com/toxicFork/react-three-renderer-example/blob/master/src/examples/AdvancedExample/index.js).

# Lower level differences:
When react 0.14 was released, they have promised better support for non-dom renderers. `react-three-renderer` is built upon that promise.

So `react-three-renderer` will be similar to `react-native`, and other non-dom rendering.

I reimplement `react-dom` (specifically, the `react-dom` of somewhere around 0.14.0, since it has chaged a bit and I need to catch up) but for `THREE.js` instead of DOM. This means essentially it should be at the same cost of using `react-dom`. 

This means I do not need to monkey-patch react so it will not affect the performance of the rest of the application for example, if you had `react-three-renderer` for a tiny canvas preview, and then unmounted it, it should have zero side effects on performance nor memory.

If react changes, the monkey patching for `react-three` may break. But since I try to reuse most of react, it means that `react-three-renderer` will also probably break.

However in the future when react separates completely from DOM (if it has not already), then `react-three-renderer` will need to refer to fewer parts of react / react-dom. This `react-three-renderer` and similar projects should hopefully be more compatible and easier to support. And I am guessing future updates to react will try to be backwards compatible with other renderers, so that should also be a small bonus.

# Additional features:

## Context-sensitive native components

For example, if you have a `material` type within a `mesh` type, that will automatically assign the material to the mesh. So that becomes:
```jsx
<mesh>
  <geometry/>
  <material color={redColor}/>
</mesh>
```
instead of:
```
<mesh geometry={geometry} material={material}/>
```
This way you can update any property of material through react rather than doing it manually. For example if you wanted to use `greenColor` instead of `redColor`.

> note: `react-three-renderer` still supports geometry / material props on `<mesh/>`, but they may be deprecated in the future.

## Resources:

If you want to reuse for example a geometry resource, you can do:
```jsx
<scene>
  <resources>
    <geometry resourceId="foo" vertices={...} ... />
  </resources>
  <mesh>
    <geometryResource resourceId="foo"/>
  </mesh>
  <mesh>
    <geometryResource resourceId="foo"/>
  </mesh>
  <mesh>
    <geometryResource resourceId="foo"/>
  </mesh>
</scene>
```

Then all three meshes point to the same geometry, and when you want to update the geometry nothing should change for the mesh.

## Modules
Currently really hard to use and probably impossible for anyone other than me to create ( will fix ), but it's in the plans for `react-three-renderer` to support modules such as [mouse input](https://github.com/toxicFork/react-three-renderer-example/blob/master/src/examples/inputs/MouseInput.js). In the mouse input example, if you do not have this module, there is no cost. If you do have the module, every frame there is a raycast into the scene in order to find which objects the mouse is hovering over, and you can easily add props such as `onMouseEnter` and they should work as expected. With `react-three` it has been implemented within `react-three` itself, through `pointerEvents`. That still works of course, but I find it beneficial to support extensibility through this. 

In the future you should be able to `npm install rtr-mouse-input` and use it as you wish, and updates to `rtr-mouse-input` to add for example touch support or add `hover` events ( rather than `enter` ) should not affect `react-three-renderer`.

## Implemented types
Currently both `react-three-renderer` and `react-three` are missing some types from `THREE.js`, but we should both work towards implementing full functionality on our libs :)

## Hot reloading
[That still works](http://giphy.com/gifs/3oEdv08aMDxw5x44h2?tc=1)

## react dev tools
[That works too, with extra benefits :)](https://www.youtube.com/watch?v=0DIpC796oic), but it was a pain in the butt to implement for me. And it's hacky. But hopefully react devs should make it easier for other people to integrate dev tools with their custom renderers.

Also there is this funny situation where if you update components every single frame, react-dev-tools event buffer fills up, because they send ALL updates through a slow and thin pipe to the extension context. But that's their problem to solve I think :)

**Comment by Izzimach** [comment]
Great, thanks for the info!

I decided to make `react-three` a minimal interface to threejs because that makes it easier for me to maintain. :smiley:  But I think a lot of people prefer having the resources managed automatically the way you've done.

In any case, I'll add a link to this project since I think it makes sense for a lot of people to use `react-three-renderer`

**Comment by justinmolinaa** [comment]
Is there currently a way to import models into a component using a <JSONLoader> component ?
